gw_eccentricity/eccDefinition.py

```diff
@@ -1184,6 +1228,180 @@ def interp_extrema(self, extrema_type="pericenters"):
+        checks.
+        """
+        return get_rational_fit(x, y,
+                                rational_fit_kwargs=self.rational_fit_kwargs,
+                                check_kwargs=False)
+
+    def rational_fit(self, x, y, description=None):
+        """Get rational fit with adaptive numerator and denominator degree.
+
+        This function ensures that the rational fit uses the optimal degree for
+        the numerator and denominator. It first checks for nonmonotonicity in the
+        fit's derivative and lowers the degrees if necessary. Only if the initial
+        degree was monotonic (no reduction needed), it attempts to increase the
+        degrees to find a higher-degree fit. If increasing the degrees causes
+        nonmonotonicity, it reverts to the previous valid monotonic fit.
+        """
+        # make sure that description is not None. A description is needed to
+        # update the optimal values of the numerator and denominator degrees
+        # used to build the final rational fit
+        if description is None:
+            raise Exception("Please provide a description for which to build a "
+                            "rational fit. `description` can not be None. For example, "
+                            "it can be 'apocenters', 'pericenters' or 'omega_gw_average'.")
+        # Set initial degrees if not already specified.
+        # The optimal degrees for the numerator and denominator changes depending on the
+        # eccentricity and duration of the waveform. We set these values by counting the
+        # number of approximate orbits and using some prior mapping between number of orbits
+        # and the optimal degrees. The mapping is done via `get_approximate_degree_for_rational_fit`.
+        # Usually this is a good starting point but not the optimal one
+        # as it varies as a function of eccentricity which we do not know apriori. The optimal
+        # value is found by doing some iterations.
+        if not (self.rational_fit_kwargs["num_degree"] and self.rational_fit_kwargs["denom_degree"]):
+            self.rational_fit_kwargs["num_degree"], self.rational_fit_kwargs["denom_degree"] \
+                = self.get_approximate_degree_for_rational_fit()
+
+        rat_fit = self.get_rat_fit(x, y)
+        t = np.arange(x[0], x[-1], self.t[1] - self.t[0])
+        # save the degrees for checks at each step of iterations for finding the optimal degrees
+        old_num_degree = self.rational_fit_kwargs["num_degree"]
+        old_denom_degree = self.rational_fit_kwargs["denom_degree"]
+
+        # use a flag to know whether the initial degrees produced monotonic
+        # time derivative of the fits or it was required to be reduced.
+        degrees_were_reduced = False
+
+        # Check for nonmonotonicity and lower degrees if needed
+        while self.check_if_domega_dt_is_nonmonotonic(t, rat_fit(t), 1.0, description):
```

Comment:
Should the function be called something other than `check_if_domega_dt_is_nonmonotonic`? Isn't the data allowed to be things other than omega?

Reply:
Ok, I have changed it to `check_if_first_derivative_is_not_strictly_monotonic`